### PR TITLE
Templating: use dashboard timerange when variables are set to refresh 'On Dashboard Load'

### LINKED ISSUE #31474: Problem with elasticsearch based variables after upgrade to 7.4.3
**What happened**:
After upgrade to Grafana 7.4.3 variables using data source Elasticsearch are distorted - "Refresh" and "Sort" seem to be re-set to default values and seem to not work on load of dashboard. Only after editing then dashboards start to work again.

**What you expected to happen**:
No such problems should happen

**How to reproduce it (as minimally and precisely as possible)**:
After upgrade to 7.4.3 people start to complain about not working dashboards based on Elasticsearch data source - all these dashboards use variables with values extracted from ES. There variables were not working and started to work only after editing them.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.4.3
- Data source type & version: Elasticsearch 6.8
- OS Grafana is installed on: Debian 9/10
- User OS & Browser: different
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by Elfo404**
**What this PR does / why we need it**:
This change makes sure we pass the current dashboard time range when a query variable is set to refresh "on dashboard load"

for more context on why this is needed see https://github.com/grafana/grafana/issues/31474#issuecomment-790687646

**Which issue(s) this PR fixes**:

Fixes #31474, #31165

**Special notes for your reviewer**:

I couldn't find tests that specifically test when/how variables are refreshed. the test file I changed seems to be "unrelated" to this, meaning it only tests actions payload and not when / how many times "variables get refreshed".

I most likely overlooked something, if someone could point me to where this is/should be tested i'd be happy to add a few more tests.

Also, I'd consider this a "feature" rather than a fix so I'm targeting `7.5.0`, should I instead target`7.4.x` and consider this a fix?

/cc. @ivanahuckova we talked about this yesterday, what's your take on this?

### CODE REVIEW

**Comment by hugohaggmark** [comment]
I think you could add some tests here https://github.com/grafana/grafana/blob/2a2a6a3bfa43457e4e51899115f992649262ce46/public/app/features/variables/query/queryRunners.test.ts

File: public/app/features/variables/pickers/OptionsPicker/actions.ts [hunk 1]
```diff
@@ -144,7 +144,7 @@ const fetchTagValues = (tagText: string): ThunkResult<Promise<string[]>> => {
 };
 
 const getTimeRange = (variable: QueryVariableModel) => {
-  if (variable.refresh === VariableRefresh.onTimeRangeChanged) {
+  if (variable.refresh === VariableRefresh.onTimeRangeChanged || variable.refresh === VariableRefresh.onDashboardLoad) {
```
Comment (hugohaggmark) [inline public/app/features/variables/pickers/OptionsPicker/actions.ts:147]: This part is going away in v8 so no need for tests here.

File: public/app/features/variables/utils.ts [hunk 2]
```diff
@@ -137,7 +137,7 @@ export function getTemplatedRegex(variable: QueryVariableModel, templateSrv = ge
 
 export function getLegacyQueryOptions(variable: QueryVariableModel, searchFilter?: string, timeSrv = getTimeSrv()) {
   const queryOptions: any = { range: undefined, variable, searchFilter };
-  if (variable.refresh === VariableRefresh.onTimeRangeChanged) {
+  if (variable.refresh === VariableRefresh.onTimeRangeChanged || variable.refresh === VariableRefresh.onDashboardLoad) {
```
Comment (hugohaggmark) [inline public/app/features/variables/utils.ts:140]: I think tests can be added to https://github.com/grafana/grafana/blob/master/public/app/features/variables/query/queryRunners.test.ts

Comment (Elfo404) [reply public/app/features/variables/utils.ts:140]: thanks! 🙏

Comment (Elfo404) [reply public/app/features/variables/utils.ts:140]: added a test case in `queryRunners.test.ts` but I'm still not sure how to properly test that a variable with `refresh = onDashboardLoad` won't get refreshed when time range changes

Comment (hugohaggmark) [reply public/app/features/variables/utils.ts:140]: For this case I think you need to look into https://github.com/grafana/grafana/blob/master/public/app/features/variables/state/onTimeRangeUpdated.test.ts

**Review by hugohaggmark — APPROVED** [review]
LGTM!

**Comment by Elfo404** [comment]
@hugohaggmark created a follow-up issue to add tests in https://github.com/grafana/grafana/issues/31799 as I had some troubles adding tests for the case mentioned in https://github.com/grafana/grafana/pull/31721#discussion_r589336809

**Pull request merged**